File: dqops/src/main/java/com/dqops/checks/column/recurring/ColumnDailyRecurringCheckCategoriesSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.checks.column.recurring;

import com.dqops.checks.AbstractRootChecksContainerSpec;
import com.dqops.checks.CheckTarget;
import com.dqops.checks.CheckTimeScale;
import com.dqops.checks.CheckType;
import com.dqops.checks.column.recurring.accuracy.ColumnAccuracyDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.anomaly.ColumnAnomalyDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.bool.ColumnBoolDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.comparison.ColumnComparisonDailyRecurringChecksSpecMap;
import com.dqops.checks.column.recurring.datatype.ColumnDatatypeDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.datetime.ColumnDatetimeDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.integrity.ColumnIntegrityDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.nulls.ColumnNullsDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.numeric.ColumnNumericDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.pii.ColumnPiiDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.schema.ColumnSchemaDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.sql.ColumnSqlDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.strings.ColumnStringsDailyRecurringChecksSpec;
import com.dqops.checks.column.recurring.uniqueness.ColumnUniquenessDailyRecurringChecksSpec;
import com.dqops.metadata.timeseries.TimeSeriesConfigurationSpec;
import com.dqops.metadata.timeseries.TimePeriodGradient;
import com.dqops.metadata.timeseries.TimeSeriesMode;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.scheduling.CheckRunRecurringScheduleGroup;
import com.dqops.metadata.sources.TableSpec;
import com.dqops.utils.serialization.IgnoreEmptyYamlSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.EqualsAndHashCode;

import java.util.Objects;

/**
 * Container of column level daily recurring checks. Contains categories of daily recurring checks.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class ColumnDailyRecurringCheckCategoriesSpec extends AbstractRootChecksContainerSpec {
    public static final ChildHierarchyNodeFieldMapImpl<ColumnDailyRecurringCheckCategoriesSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractRootChecksContainerSpec.FIELDS) {
        {
            put("nulls", o -> o.nulls);
            put("numeric", o -> o.numeric);
            put("strings", o -> o.strings);
            put("uniqueness", o -> o.uniqueness);
            put("datetime", o -> o.datetime);
            put("pii", o -> o.pii);
            put("sql", o -> o.sql);
            put("bool", o -> o.bool);
            put("integrity", o -> o.integrity);
            put("accuracy", o -> o.accuracy);
            put("datatype", o -> o.datatype);
            put("anomaly", o -> o.anomaly);
            put("schema", o -> o.schema);
            put("comparisons", o -> o.comparisons);
        }
    };

    @JsonPropertyDescription("Daily recurring checks of nulls in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnNullsDailyRecurringChecksSpec nulls;

    @JsonPropertyDescription("Daily recurring checks of numeric in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnNumericDailyRecurringChecksSpec numeric;

    @JsonPropertyDescription("Daily recurring checks of strings in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnStringsDailyRecurringChecksSpec strings;

    @JsonPropertyDescription("Daily recurring checks of uniqueness in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnUniquenessDailyRecurringChecksSpec uniqueness;

    @JsonPropertyDescription("Daily recurring checks of datetime in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnDatetimeDailyRecurringChecksSpec datetime;

    @JsonPropertyDescription("Daily recurring checks of Personal Identifiable Information (PII) in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnPiiDailyRecurringChecksSpec pii;

    @JsonPropertyDescription("Daily recurring checks of custom SQL checks in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnSqlDailyRecurringChecksSpec sql;

    @JsonPropertyDescription("Daily recurring checks of booleans in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnBoolDailyRecurringChecksSpec bool;

    @JsonPropertyDescription("Daily recurring checks of integrity in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnIntegrityDailyRecurringChecksSpec integrity;

    @JsonPropertyDescription("Daily recurring checks of accuracy in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnAccuracyDailyRecurringChecksSpec accuracy;

    @JsonPropertyDescription("Daily recurring checks of datatype in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnDatatypeDailyRecurringChecksSpec datatype;

    @JsonPropertyDescription("Daily recurring checks of anomaly in the column")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnAnomalyDailyRecurringChecksSpec anomaly;

    @JsonPropertyDescription("Daily recurring column schema checks")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnSchemaDailyRecurringChecksSpec schema;

    @JsonPropertyDescription("Dictionary of configuration of checks for table comparisons at a column level. The key that identifies each comparison must match the name of a data comparison that is configured on the parent table.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private ColumnComparisonDailyRecurringChecksSpecMap comparisons = new ColumnComparisonDailyRecurringChecksSpecMap();

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnNullsDailyRecurringChecksSpec getNulls() {
        return nulls;
    }

    /**
     * Sets the container of nulls data quality checks (recurring).
     * @param nulls New nulls checks.
     */
    public void setNulls(ColumnNullsDailyRecurringChecksSpec nulls) {
		this.setDirtyIf(!Objects.equals(this.nulls, nulls));
        this.nulls = nulls;
		this.propagateHierarchyIdToField(nulls, "nulls");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnNumericDailyRecurringChecksSpec getNumeric() {
        return numeric;
    }

    /**
     * Sets the container of numeric data quality checks (recurring).
     * @param numeric New numeric checks.
     */
    public void setNumeric(ColumnNumericDailyRecurringChecksSpec numeric) {
        this.setDirtyIf(!Objects.equals(this.numeric, numeric));
        this.numeric = numeric;
        this.propagateHierarchyIdToField(numeric, "numeric");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnStringsDailyRecurringChecksSpec getStrings() {
        return strings;
    }

    /**
     * Sets the container of strings data quality checks (recurring).
     * @param strings New strings checks.
     */
    public void setStrings(ColumnStringsDailyRecurringChecksSpec strings) {
        this.setDirtyIf(!Objects.equals(this.strings, strings));
        this.strings = strings;
        this.propagateHierarchyIdToField(strings, "strings");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnUniquenessDailyRecurringChecksSpec getUniqueness() {
        return uniqueness;
    }

    /**
     * Sets the container of uniqueness data quality checks (recurring).
     * @param uniqueness New uniqueness checks.
     */
    public void setUniqueness(ColumnUniquenessDailyRecurringChecksSpec uniqueness) {
        this.setDirtyIf(!Objects.equals(this.uniqueness, uniqueness));
        this.uniqueness = uniqueness;
        this.propagateHierarchyIdToField(uniqueness, "uniqueness");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnDatetimeDailyRecurringChecksSpec getDatetime() {
        return datetime;
    }

    /**
     * Sets the container of datetime data quality checks (recurring).
     * @param datetime New datetime checks.
     */
    public void setDatetime(ColumnDatetimeDailyRecurringChecksSpec datetime) {
        this.setDirtyIf(!Objects.equals(this.datetime, datetime));
        this.datetime = datetime;
        this.propagateHierarchyIdToField(datetime, "datetime");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnPiiDailyRecurringChecksSpec getPii() {
        return pii;
    }

    /**
     * Sets the container of Personal Identifiable Information (PII) data quality checks (recurring).
     * @param pii New Personal Identifiable Information (PII) checks.
     */
    public void setPii(ColumnPiiDailyRecurringChecksSpec pii) {
        this.setDirtyIf(!Objects.equals(this.pii, pii));
        this.pii = pii;
        this.propagateHierarchyIdToField(pii, "pii");
    }

    /**
     * Returns the container of custom SQL checks that use custom SQL expressions in checks.
     * @return Custom SQL checks.
     */
    public ColumnSqlDailyRecurringChecksSpec getSql() {
        return sql;
    }

    /**
     * Sets a reference to a container of custom SQL checks.
     * @param sql Custom SQL checks.
     */
    public void setSql(ColumnSqlDailyRecurringChecksSpec sql) {
        this.setDirtyIf(!Objects.equals(this.sql, sql));
        this.sql = sql;
        this.propagateHierarchyIdToField(sql, "sql");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnBoolDailyRecurringChecksSpec getBool() {
        return bool;
    }

    /**
     * Sets the container of booleans data quality checks (recurring).
     * @param bool New booleans checks.
     */
    public void setBool(ColumnBoolDailyRecurringChecksSpec bool) {
        this.setDirtyIf(!Objects.equals(this.bool, bool));
        this.bool = bool;
        this.propagateHierarchyIdToField(bool, "bool");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnIntegrityDailyRecurringChecksSpec getIntegrity() {
        return integrity;
    }

    /**
     * Sets the container of integrity data quality checks (recurring).
     * @param integrity New integrity checks.
     */
    public void setIntegrity(ColumnIntegrityDailyRecurringChecksSpec integrity) {
        this.setDirtyIf(!Objects.equals(this.integrity, integrity));
        this.integrity = integrity;
        this.propagateHierarchyIdToField(integrity, "integrity");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnAccuracyDailyRecurringChecksSpec getAccuracy() {
        return accuracy;
    }

    /**
     * Sets the container of accuracy data quality checks (recurring).
     * @param accuracy New accuracy checks.
     */
    public void setAccuracy(ColumnAccuracyDailyRecurringChecksSpec accuracy) {
        this.setDirtyIf(!Objects.equals(this.accuracy, accuracy));
        this.accuracy = accuracy;
        this.propagateHierarchyIdToField(accuracy, "accuracy");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of row standard data quality recurring.
     */
    public ColumnDatatypeDailyRecurringChecksSpec getDatatype() {
        return datatype;
    }

    /**
     * Sets the container of datatype data quality checks (recurring).
     * @param datatype New datatype checks.
     */
    public void setDatatype(ColumnDatatypeDailyRecurringChecksSpec datatype) {
        this.setDirtyIf(!Objects.equals(this.datatype, datatype));
        this.datatype = datatype;
        this.propagateHierarchyIdToField(datatype, "datatype");
    }

    /**
     * Returns the container of recurring for standard data quality checks.
     * @return Container of anomaly data quality recurring.
     */
    public ColumnAnomalyDailyRecurringChecksSpec getAnomaly() {
        return anomaly;
    }

    /**
     * Sets the container of anomaly data quality checks (recurring).
     * @param anomaly New anomaly checks.
     */
    public void setAnomaly(ColumnAnomalyDailyRecurringChecksSpec anomaly) {
        this.setDirtyIf(!Objects.equals(this.anomaly, anomaly));
        this.anomaly = anomaly;
        this.propagateHierarchyIdToField(anomaly, "anomaly");
    }

    /**
     * Returns the container of daily recurring column schema checks.
     * @return Container of column schema checks.
     */
    public ColumnSchemaDailyRecurringChecksSpec getSchema() {
        return schema;
    }

    /**
     * Sets the container of daily recurring schema checks.
     * @param schema Container of schema checks.
     */
    public void setSchema(ColumnSchemaDailyRecurringChecksSpec schema) {
        this.setDirtyIf(!Objects.equals(this.schema, schema));
        this.schema = schema;
        this.propagateHierarchyIdToField(schema, "schema");
    }

    /**
     * Returns the container of column level comparisons to columns in the reference table.
     * @return Dictionary of comparisons to columns.
     */
    @Override
    public ColumnComparisonDailyRecurringChecksSpecMap getComparisons() {
        return comparisons;
    }

    /**
     * Sets the container of named comparisons to columns in other reference tables.
     * @param comparisons Container of column level comparisons.
     */
    public void setComparisons(ColumnComparisonDailyRecurringChecksSpecMap comparisons) {
        this.setDirtyIf(!Objects.equals(this.comparisons, comparisons));
        this.comparisons = comparisons;
        this.propagateHierarchyIdToField(comparisons, "comparisons");
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Returns time series configuration for the given group of checks.
     *
     * @param tableSpec Parent table specification - used to get the details about the time partitioning column.
     * @return Time series configuration.
     */
    @Override
    public TimeSeriesConfigurationSpec getTimeSeriesConfiguration(TableSpec tableSpec) {
        return new TimeSeriesConfigurationSpec()
        {{
            setMode(TimeSeriesMode.current_time);
            setTimeGradient(TimePeriodGradient.day);
        }};
    }

    /**
     * Returns the type of checks (profiling, recurring, partitioned).
     *
     * @return Check type.
     */
    @Override
    @JsonIgnore
    public CheckType getCheckType() {
        return CheckType.recurring;
    }

    /**
     * Returns the time range for recurring and partitioned checks (daily, monthly, etc.).
     * Profiling checks do not have a time range and return null.
     *
     * @return Time range (daily, monthly, ...).
     */
    @Override
    @JsonIgnore
    public CheckTimeScale getCheckTimeScale() {
        return CheckTimeScale.daily;
    }

    /**
     * Returns the check target, where the check could be applied.
     *
     * @return Check target, "table" or "column".
     */
    @Override
    @JsonIgnore
    public CheckTarget getCheckTarget() {
        return CheckTarget.column;
    }

    /**
     * Returns the name of the cron expression that is used to schedule checks in this check root object.
     *
     * @return Recurring schedule group (named schedule) that is used to schedule the checks in this root.
     */
    @Override
    @JsonIgnore
    public CheckRunRecurringScheduleGroup getSchedulingGroup() {
        return CheckRunRecurringScheduleGroup.recurring_daily;
    }
}

File: distribution/python/dqops/client/models/column_daily_recurring_check_categories_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

import attr

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.column_accuracy_daily_recurring_checks_spec import (
        ColumnAccuracyDailyRecurringChecksSpec,
    )
    from ..models.column_anomaly_daily_recurring_checks_spec import (
        ColumnAnomalyDailyRecurringChecksSpec,
    )
    from ..models.column_bool_daily_recurring_checks_spec import (
        ColumnBoolDailyRecurringChecksSpec,
    )
    from ..models.column_daily_recurring_check_categories_spec_comparisons import (
        ColumnDailyRecurringCheckCategoriesSpecComparisons,
    )
    from ..models.column_daily_recurring_check_categories_spec_custom import (
        ColumnDailyRecurringCheckCategoriesSpecCustom,
    )
    from ..models.column_datatype_daily_recurring_checks_spec import (
        ColumnDatatypeDailyRecurringChecksSpec,
    )
    from ..models.column_datetime_daily_recurring_checks_spec import (
        ColumnDatetimeDailyRecurringChecksSpec,
    )
    from ..models.column_integrity_daily_recurring_checks_spec import (
        ColumnIntegrityDailyRecurringChecksSpec,
    )
    from ..models.column_nulls_daily_recurring_checks_spec import (
        ColumnNullsDailyRecurringChecksSpec,
    )
    from ..models.column_numeric_daily_recurring_checks_spec import (
        ColumnNumericDailyRecurringChecksSpec,
    )
    from ..models.column_pii_daily_recurring_checks_spec import (
        ColumnPiiDailyRecurringChecksSpec,
    )
    from ..models.column_schema_daily_recurring_checks_spec import (
        ColumnSchemaDailyRecurringChecksSpec,
    )
    from ..models.column_sql_daily_recurring_checks_spec import (
        ColumnSqlDailyRecurringChecksSpec,
    )
    from ..models.column_strings_daily_recurring_checks_spec import (
        ColumnStringsDailyRecurringChecksSpec,
    )
    from ..models.column_uniqueness_daily_recurring_checks_spec import (
        ColumnUniquenessDailyRecurringChecksSpec,
    )


T = TypeVar("T", bound="ColumnDailyRecurringCheckCategoriesSpec")


@attr.s(auto_attribs=True)
class ColumnDailyRecurringCheckCategoriesSpec:
    """
    Attributes:
        custom (Union[Unset, ColumnDailyRecurringCheckCategoriesSpecCustom]): Dictionary of custom checks. The keys are
            check names.
        nulls (Union[Unset, ColumnNullsDailyRecurringChecksSpec]):
        numeric (Union[Unset, ColumnNumericDailyRecurringChecksSpec]):
        strings (Union[Unset, ColumnStringsDailyRecurringChecksSpec]):
        uniqueness (Union[Unset, ColumnUniquenessDailyRecurringChecksSpec]):
        datetime_ (Union[Unset, ColumnDatetimeDailyRecurringChecksSpec]):
        pii (Union[Unset, ColumnPiiDailyRecurringChecksSpec]):
        sql (Union[Unset, ColumnSqlDailyRecurringChecksSpec]):
        bool_ (Union[Unset, ColumnBoolDailyRecurringChecksSpec]):
        integrity (Union[Unset, ColumnIntegrityDailyRecurringChecksSpec]):
        accuracy (Union[Unset, ColumnAccuracyDailyRecurringChecksSpec]):
        datatype (Union[Unset, ColumnDatatypeDailyRecurringChecksSpec]):
        anomaly (Union[Unset, ColumnAnomalyDailyRecurringChecksSpec]):
        schema (Union[Unset, ColumnSchemaDailyRecurringChecksSpec]):
        comparisons (Union[Unset, ColumnDailyRecurringCheckCategoriesSpecComparisons]): Dictionary of configuration of
            checks for table comparisons at a column level. The key that identifies each comparison must match the name of a
            data comparison that is configured on the parent table.
    """

    custom: Union[Unset, "ColumnDailyRecurringCheckCategoriesSpecCustom"] = UNSET
    nulls: Union[Unset, "ColumnNullsDailyRecurringChecksSpec"] = UNSET
    numeric: Union[Unset, "ColumnNumericDailyRecurringChecksSpec"] = UNSET
    strings: Union[Unset, "ColumnStringsDailyRecurringChecksSpec"] = UNSET
    uniqueness: Union[Unset, "ColumnUniquenessDailyRecurringChecksSpec"] = UNSET
    datetime_: Union[Unset, "ColumnDatetimeDailyRecurringChecksSpec"] = UNSET
    pii: Union[Unset, "ColumnPiiDailyRecurringChecksSpec"] = UNSET
    sql: Union[Unset, "ColumnSqlDailyRecurringChecksSpec"] = UNSET
    bool_: Union[Unset, "ColumnBoolDailyRecurringChecksSpec"] = UNSET
    integrity: Union[Unset, "ColumnIntegrityDailyRecurringChecksSpec"] = UNSET
    accuracy: Union[Unset, "ColumnAccuracyDailyRecurringChecksSpec"] = UNSET
    datatype: Union[Unset, "ColumnDatatypeDailyRecurringChecksSpec"] = UNSET
    anomaly: Union[Unset, "ColumnAnomalyDailyRecurringChecksSpec"] = UNSET
    schema: Union[Unset, "ColumnSchemaDailyRecurringChecksSpec"] = UNSET
    comparisons: Union[
        Unset, "ColumnDailyRecurringCheckCategoriesSpecComparisons"
    ] = UNSET
    additional_properties: Dict[str, Any] = attr.ib(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        custom: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.custom, Unset):
            custom = self.custom.to_dict()

        nulls: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.nulls, Unset):
            nulls = self.nulls.to_dict()

        numeric: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.numeric, Unset):
            numeric = self.numeric.to_dict()

        strings: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.strings, Unset):
            strings = self.strings.to_dict()

        uniqueness: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.uniqueness, Unset):
            uniqueness = self.uniqueness.to_dict()

        datetime_: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.datetime_, Unset):
            datetime_ = self.datetime_.to_dict()

        pii: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.pii, Unset):
            pii = self.pii.to_dict()

        sql: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.sql, Unset):
            sql = self.sql.to_dict()

        bool_: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.bool_, Unset):
            bool_ = self.bool_.to_dict()

        integrity: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.integrity, Unset):
            integrity = self.integrity.to_dict()

        accuracy: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.accuracy, Unset):
            accuracy = self.accuracy.to_dict()

        datatype: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.datatype, Unset):
            datatype = self.datatype.to_dict()

        anomaly: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.anomaly, Unset):
            anomaly = self.anomaly.to_dict()

        schema: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.schema, Unset):
            schema = self.schema.to_dict()

        comparisons: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.comparisons, Unset):
            comparisons = self.comparisons.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if custom is not UNSET:
            field_dict["custom"] = custom
        if nulls is not UNSET:
            field_dict["nulls"] = nulls
        if numeric is not UNSET:
            field_dict["numeric"] = numeric
        if strings is not UNSET:
            field_dict["strings"] = strings
        if uniqueness is not UNSET:
            field_dict["uniqueness"] = uniqueness
        if datetime_ is not UNSET:
            field_dict["datetime"] = datetime_
        if pii is not UNSET:
            field_dict["pii"] = pii
        if sql is not UNSET:
            field_dict["sql"] = sql
        if bool_ is not UNSET:
            field_dict["bool"] = bool_
        if integrity is not UNSET:
            field_dict["integrity"] = integrity
        if accuracy is not UNSET:
            field_dict["accuracy"] = accuracy
        if datatype is not UNSET:
            field_dict["datatype"] = datatype
        if anomaly is not UNSET:
            field_dict["anomaly"] = anomaly
        if schema is not UNSET:
            field_dict["schema"] = schema
        if comparisons is not UNSET:
            field_dict["comparisons"] = comparisons

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.column_accuracy_daily_recurring_checks_spec import (
            ColumnAccuracyDailyRecurringChecksSpec,
        )
        from ..models.column_anomaly_daily_recurring_checks_spec import (
            ColumnAnomalyDailyRecurringChecksSpec,
        )
        from ..models.column_bool_daily_recurring_checks_spec import (
            ColumnBoolDailyRecurringChecksSpec,
        )
        from ..models.column_daily_recurring_check_categories_spec_comparisons import (
            ColumnDailyRecurringCheckCategoriesSpecComparisons,
        )
        from ..models.column_daily_recurring_check_categories_spec_custom import (
            ColumnDailyRecurringCheckCategoriesSpecCustom,
        )
        from ..models.column_datatype_daily_recurring_checks_spec import (
            ColumnDatatypeDailyRecurringChecksSpec,
        )
        from ..models.column_datetime_daily_recurring_checks_spec import (
            ColumnDatetimeDailyRecurringChecksSpec,
        )
        from ..models.column_integrity_daily_recurring_checks_spec import (
            ColumnIntegrityDailyRecurringChecksSpec,
        )
        from ..models.column_nulls_daily_recurring_checks_spec import (
            ColumnNullsDailyRecurringChecksSpec,
        )
        from ..models.column_numeric_daily_recurring_checks_spec import (
            ColumnNumericDailyRecurringChecksSpec,
        )
        from ..models.column_pii_daily_recurring_checks_spec import (
            ColumnPiiDailyRecurringChecksSpec,
        )
        from ..models.column_schema_daily_recurring_checks_spec import (
            ColumnSchemaDailyRecurringChecksSpec,
        )
        from ..models.column_sql_daily_recurring_checks_spec import (
            ColumnSqlDailyRecurringChecksSpec,
        )
        from ..models.column_strings_daily_recurring_checks_spec import (
            ColumnStringsDailyRecurringChecksSpec,
        )
        from ..models.column_uniqueness_daily_recurring_checks_spec import (
            ColumnUniquenessDailyRecurringChecksSpec,
        )

        d = src_dict.copy()
        _custom = d.pop("custom", UNSET)
        custom: Union[Unset, ColumnDailyRecurringCheckCategoriesSpecCustom]
        if isinstance(_custom, Unset):
            custom = UNSET
        else:
            custom = ColumnDailyRecurringCheckCategoriesSpecCustom.from_dict(_custom)

        _nulls = d.pop("nulls", UNSET)
        nulls: Union[Unset, ColumnNullsDailyRecurringChecksSpec]
        if isinstance(_nulls, Unset):
            nulls = UNSET
        else:
            nulls = ColumnNullsDailyRecurringChecksSpec.from_dict(_nulls)

        _numeric = d.pop("numeric", UNSET)
        numeric: Union[Unset, ColumnNumericDailyRecurringChecksSpec]
        if isinstance(_numeric, Unset):
            numeric = UNSET
        else:
            numeric = ColumnNumericDailyRecurringChecksSpec.from_dict(_numeric)

        _strings = d.pop("strings", UNSET)
        strings: Union[Unset, ColumnStringsDailyRecurringChecksSpec]
        if isinstance(_strings, Unset):
            strings = UNSET
        else:
            strings = ColumnStringsDailyRecurringChecksSpec.from_dict(_strings)

        _uniqueness = d.pop("uniqueness", UNSET)
        uniqueness: Union[Unset, ColumnUniquenessDailyRecurringChecksSpec]
        if isinstance(_uniqueness, Unset):
            uniqueness = UNSET
        else:
            uniqueness = ColumnUniquenessDailyRecurringChecksSpec.from_dict(_uniqueness)

        _datetime_ = d.pop("datetime", UNSET)
        datetime_: Union[Unset, ColumnDatetimeDailyRecurringChecksSpec]
        if isinstance(_datetime_, Unset):
            datetime_ = UNSET
        else:
            datetime_ = ColumnDatetimeDailyRecurringChecksSpec.from_dict(_datetime_)

        _pii = d.pop("pii", UNSET)
        pii: Union[Unset, ColumnPiiDailyRecurringChecksSpec]
        if isinstance(_pii, Unset):
            pii = UNSET
        else:
            pii = ColumnPiiDailyRecurringChecksSpec.from_dict(_pii)

        _sql = d.pop("sql", UNSET)
        sql: Union[Unset, ColumnSqlDailyRecurringChecksSpec]
        if isinstance(_sql, Unset):
            sql = UNSET
        else:
            sql = ColumnSqlDailyRecurringChecksSpec.from_dict(_sql)

        _bool_ = d.pop("bool", UNSET)
        bool_: Union[Unset, ColumnBoolDailyRecurringChecksSpec]
        if isinstance(_bool_, Unset):
            bool_ = UNSET
        else:
            bool_ = ColumnBoolDailyRecurringChecksSpec.from_dict(_bool_)

        _integrity = d.pop("integrity", UNSET)
        integrity: Union[Unset, ColumnIntegrityDailyRecurringChecksSpec]
        if isinstance(_integrity, Unset):
            integrity = UNSET
        else:
            integrity = ColumnIntegrityDailyRecurringChecksSpec.from_dict(_integrity)

        _accuracy = d.pop("accuracy", UNSET)
        accuracy: Union[Unset, ColumnAccuracyDailyRecurringChecksSpec]
        if isinstance(_accuracy, Unset):
            accuracy = UNSET
        else:
            accuracy = ColumnAccuracyDailyRecurringChecksSpec.from_dict(_accuracy)

        _datatype = d.pop("datatype", UNSET)
        datatype: Union[Unset, ColumnDatatypeDailyRecurringChecksSpec]
        if isinstance(_datatype, Unset):
            datatype = UNSET
        else:
            datatype = ColumnDatatypeDailyRecurringChecksSpec.from_dict(_datatype)

        _anomaly = d.pop("anomaly", UNSET)
        anomaly: Union[Unset, ColumnAnomalyDailyRecurringChecksSpec]
        if isinstance(_anomaly, Unset):
            anomaly = UNSET
        else:
            anomaly = ColumnAnomalyDailyRecurringChecksSpec.from_dict(_anomaly)

        _schema = d.pop("schema", UNSET)
        schema: Union[Unset, ColumnSchemaDailyRecurringChecksSpec]
        if isinstance(_schema, Unset):
            schema = UNSET
        else:
            schema = ColumnSchemaDailyRecurringChecksSpec.from_dict(_schema)

        _comparisons = d.pop("comparisons", UNSET)
        comparisons: Union[Unset, ColumnDailyRecurringCheckCategoriesSpecComparisons]
        if isinstance(_comparisons, Unset):
            comparisons = UNSET
        else:
            comparisons = ColumnDailyRecurringCheckCategoriesSpecComparisons.from_dict(
                _comparisons
            )

        column_daily_recurring_check_categories_spec = cls(
            custom=custom,
            nulls=nulls,
            numeric=numeric,
            strings=strings,
            uniqueness=uniqueness,
            datetime_=datetime_,
            pii=pii,
            sql=sql,
            bool_=bool_,
            integrity=integrity,
            accuracy=accuracy,
            datatype=datatype,
            anomaly=anomaly,
            schema=schema,
            comparisons=comparisons,
        )

        column_daily_recurring_check_categories_spec.additional_properties = d
        return column_daily_recurring_check_categories_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
